docs/source/jit_type_system_1.rst

```diff
@@ -0,0 +1,628 @@
+
+    def fn(a: MyClass):
+        return a.name
+
+It leads to the following compile-time error:
+
+.. testoutput::
+
+    RuntimeError:
+    Tried to access nonexistent attribute or method 'name' of type '__torch__.MyClass'. Did you forget to initialize an attribute in __init__()?:
+        File "test-class2.py", line 10
+    def fn(a: MyClass):
+        return a.name
+            ~~~~~~ <--- HERE
+
+Enum Type
+^^^^^^^^^^
+
+Like custom classes, semantics of enum type are user-defined and the entire class definition must be compilable to TorchScript IR and subject to TorchScript type-checking rules.
+
+::
+
+    TSEnumDef := "class" Identifier "(enum.Enum | TSEnumType)" ":"
+                ( MemberIdentifier "=" Value )+
+                ( MethodDefinition )*
+
+where
+
+* Value must be TorchScript literals of type ``int``, ``float``, or ``str``, and must be of the same TorchScript type
+* ``TSEnumType`` is the name of a TorchScript ``enum`` type. Similar to Python enum, TorchScript allows restricted Enum subclassing, that is, subclassing an enum is allowed only if the enum does not define any members.
+
+Compared to Python
+""""""""""""""""""
+
+* TorchScript supports only ``enum.Enum``, but not other variations such as ``enum.IntEnum``, ``enum.Flag``, ``enum.IntFlag``, or  ``enum.auto``
+* Values of TorchScript enum members must be of the same type and can only be of ``int``, ``float``, or ``str`` type, whereas Python enum members can be of any type
+* Enums containing methods, are ignored in TorchScript.
+
+Example:
+
+.. testcode::
+
+    import torch
+    from enum import Enum
+
+    class Color(Enum):
+        RED = 1
+        GREEN = 2
+
+    def enum_fn(x: Color, y: Color) -> bool:
+        if x == Color.RED:
+            return True
+        return x == y
+
+    m = torch.jit.script(enum_fn)
+
+    print("Eager: ", enum_fn(Color.RED, Color.GREEN))
+    print("TorchScript: ", m(Color.RED, Color.GREEN))
+
+Example:
+
+The following example shows the case of restricted enum subclassing, where BaseColor does not define any member, thus can be subclassed by Color.
+
+.. testcode::
+
+    import torch
+    from enum import Enum
+
+    class BaseColor(Enum):
+        def foo(self):
+            pass
+
+    class Color(BaseColor):
+        RED = 1
+        GREEN = 2
+
+    def enum_fn(x: Color, y: Color) -> bool:
+        if x == Color.RED:
+            return True
+
+        return x == y
+
+    m = torch.jit.script(enum_fn)
+    print("TorchScript: ", m(Color.RED, Color.GREEN))
+    print("Eager: ", enum_fn(Color.RED, Color.GREEN))
+
+TorchScript Module Class
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+``TSModuleType`` is a special class type that is inferred from object instances created outside TorchScript. Although ``TSModuleType`` is named by the Python class of the object instance. Method ``__init__()`` of the Python class are not considered as a TorchScript method, thus it does not have to comply to TorchScript’s type checking rules.
```

Comment:
```suggestion
``TSModuleType`` is a special class type that is inferred from object instances created outside TorchScript. Although ``TSModuleType`` is named by the Python class of the object instance. The ``__init__()`` method of the Python class is not considered as a TorchScript method, so it does not have to comply with TorchScript’s type checking rules.
```
"Although ``TSModuleType`` is named by the Python class of the object instance" is also an incomplete sentence.